src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -131,6 +141,55 @@ internal void Write(byte[] buffer, int offset, int count)
             ProcessWrite(buffer, offset, count, null);
         }
 
+        internal Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (!_sslState.InnerStream.CanWrite) return Task.FromException(new NotSupportedException(SR.NotSupported_UnwritableStream));
+            // If cancellation was requested, bail early with an already completed task.
+            // Otherwise, return a task that represents the Begin/End methods.
+            return cancellationToken.IsCancellationRequested
+                        ? Task.FromCanceled(cancellationToken)
+                        : WriteAsyncImpl(buffer, offset, count, cancellationToken);
+        }
+
+        private async Task WriteAsyncImpl(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            var semaphore = EnsureAsyncActiveWriteSemaphoreInitialized();
+            await semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+            InitiateWrite(buffer, offset, count, null);
```

Comment:
Previously wouldn't any exceptions from InitiateWrite (e.g. another operation in progress) have emerged synchronously from the call to WriteAsync?  Now they're going to be stored into the returned task.  Following my other comments, this InitiateWrite call can just be moved to the WriteAsync caller.